homeassistant/components/light/tradfri.py

```diff
@@ -4,51 +4,80 @@
 For more details about this platform, please refer to the documentation at
 https://home-assistant.io/components/light.tradfri/
 """
+import asyncio
 import logging
 
+try:
+    from asyncio import ensure_future
+except ImportError:
+    from asyncio import async as ensure_future
+
 from homeassistant.components.light import (
     ATTR_BRIGHTNESS, ATTR_COLOR_TEMP, ATTR_RGB_COLOR, SUPPORT_BRIGHTNESS,
     SUPPORT_COLOR_TEMP, SUPPORT_RGB_COLOR, Light)
 from homeassistant.components.light import \
     PLATFORM_SCHEMA as LIGHT_PLATFORM_SCHEMA
-from homeassistant.components.tradfri import KEY_GATEWAY, KEY_TRADFRI_GROUPS
+from homeassistant.components.tradfri import KEY_GATEWAY, KEY_TRADFRI_GROUPS, \
+    KEY_API
 from homeassistant.util import color as color_util
 
 _LOGGER = logging.getLogger(__name__)
 
 DEPENDENCIES = ['tradfri']
 PLATFORM_SCHEMA = LIGHT_PLATFORM_SCHEMA
 IKEA = 'IKEA of Sweden'
+TRADFRI_LIGHT_MANAGER = 'Tradfri Light Manager'
 ALLOWED_TEMPERATURES = {
     IKEA: {2200: 'efd275', 2700: 'f1e0b5', 4000: 'f5faf6'}
 }
 
 
-def setup_platform(hass, config, add_devices, discovery_info=None):
+@asyncio.coroutine
+def async_setup_platform(hass, config, add_devices, discovery_info=None):
     """Set up the IKEA Tradfri Light platform."""
     if discovery_info is None:
         return
 
     gateway_id = discovery_info['gateway']
     gateway = hass.data[KEY_GATEWAY][gateway_id]
-    devices = gateway.get_devices()
+    api = hass.data[KEY_API]
+
+    devices_command = gateway.get_devices()
+    devices_commands = yield from api(devices_command)
+    devices = yield from api(*devices_commands)
     lights = [dev for dev in devices if dev.has_light_control]
-    add_devices(Tradfri(light) for light in lights)
+    add_devices(TradfriLight(light, api, hass) for light in lights)
 
     allow_tradfri_groups = hass.data[KEY_TRADFRI_GROUPS][gateway_id]
     if allow_tradfri_groups:
-        groups = gateway.get_groups()
-        add_devices(TradfriGroup(group) for group in groups)
+        groups_command = gateway.get_groups()
+        groups_commands = yield from api(groups_command)
+        groups = yield from api(*groups_commands)
+        add_devices(TradfriGroup(group, api, hass) for group in groups)
 
 
 class TradfriGroup(Light):
     """The platform class required by hass."""
 
-    def __init__(self, light):
+    def __init__(self, light, api, hass):
         """Initialize a Group."""
+        self._hass = hass
```

Comment:
`hass` will be set on the entity when added, so you don't need to pass `hass` in, you can just access it as an  instance attribute.